internal/encryption/keyprovider.go

```diff
@@ -88,76 +91,57 @@ func (e *targetBuilder) setupKeyProvider(cfg config.KeyProviderConfig, stack []c
 	keyProviderDescriptor, err := e.reg.GetKeyProviderDescriptor(id)
 	if err != nil {
 		if errors.Is(err, &registry.KeyProviderNotFoundError{}) {
-			return hcl.Diagnostics{&hcl.Diagnostic{
+			return diags.Append(&hcl.Diagnostic{
 				Severity: hcl.DiagError,
 				Summary:  "Unknown key_provider type",
 				Detail:   fmt.Sprintf("Can not find %q", cfg.Type),
-			}}
+			})
 		}
-		return hcl.Diagnostics{&hcl.Diagnostic{
+		return diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,
 			Summary:  fmt.Sprintf("Error fetching key_provider %q", cfg.Type),
 			Detail:   err.Error(),
-		}}
+		})
 	}
 
 	// Now that we know we have the correct Descriptor, we can decode the configuration
 	// and build the KeyProvider
 	keyProviderConfig := keyProviderDescriptor.ConfigStruct()
 
 	// Locate all the dependencies
-	deps, diags := gohcl.VariablesInBody(cfg.Body, keyProviderConfig)
+	deps, varDiags := gohcl.VariablesInBody(cfg.Body, keyProviderConfig)
+	diags = append(diags, varDiags...)
 	if diags.HasErrors() {
 		return diags
 	}
 
-	// Required Dependencies
-	for _, dep := range deps {
-		// Key Provider references should be in the form key_provider.type.name
-		if len(dep) != 3 {
-			diags = append(diags, &hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid key_provider reference",
-				Detail:   "Expected reference in form key_provider.type.name",
-				Subject:  dep.SourceRange().Ptr(),
-			})
-			continue
-		}
-
-		// TODO this should be more defensive
-		depRoot := (dep[0].(hcl.TraverseRoot)).Name
-		depType := (dep[1].(hcl.TraverseAttr)).Name
-		depName := (dep[2].(hcl.TraverseAttr)).Name
-
-		if depRoot != "key_provider" {
-			diags = append(diags, &hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid key_provider reference",
-				Detail:   "Expected reference in form key_provider.type.name",
-				Subject:  dep.SourceRange().Ptr(),
-			})
-			continue
-		}
-
-		for _, kpc := range e.cfg.KeyProviderConfigs {
-			// Find the key provider in the config
-			if kpc.Type == depType && kpc.Name == depName {
-				depDiags := e.setupKeyProvider(kpc, stack)
-				diags = append(diags, depDiags...)
-				break
-			}
-		}
+	refs, refDiags := lang.References(addrs.ParseRef, deps)
```

Comment:
This removes the unused functionality that a key_provider can depend on another key provider?

Reply:
No, I just removed the check that validates all the deps are refs to key providers.

If we want to remove the code relevant to provider chaining, the refactoring should be done in a separate PR.

Reply:
Oh, I missed the line where we call setupKeyProvider. I'll get it back slightly changing the validation logic

Reply:
I fixed this accidental removal in 2766e27

Reply:
I think you need to filter out deps that are not key_providers here.  lang.References will probably break on those traversals.

Reply:
Added in [0a61405](https://github.com/opentofu/opentofu/pull/1728/commits/0a61405e6b96fbe5db7c9e717678ae96625b9f99)